src/models/geograph.py

```diff
@@ -7,25 +7,29 @@
 
 import bz2
 import gzip
+import inspect
 import os
 import pathlib
 import pickle
 from copy import deepcopy
-from dataclasses import dataclass
 from itertools import zip_longest
-from typing import Dict, List, Optional, Tuple, Union
+from typing import Any, Callable, Dict, List, Optional, Union
 
 import geopandas as gpd
 import networkx as nx
 import numpy as np
+import pandas as pd
 import pyproj
 import rasterio
 import shapely
 from shapely.prepared import prep
 from src.data_loading import rasterio_utils
-from src.models import binary_graph_operations
+from src.models import binary_graph_operations, metrics
+from src.models.metrics import Metric
 from tqdm import tqdm
 
+pd.options.mode.chained_assignment = None  # default='warn'
```

Comment:
This is because in the `add_habitat` function, we create a new habitat by slicing the main df to get the rows which correspond to the habitat. Further operations on the habitat df then give a Pandas warning about chained assignment, which is explained here in the first answer: https://stackoverflow.com/questions/20625582/how-to-deal-with-settingwithcopywarning-in-pandas, but for our use case we don't care about it so I disabled the warning.